### PR TITLE
[Filter] Saving filter histories

### PR DESCRIPTION
**Pull request by kayla-kwon**
- Filter search history will be saved in the MongoDB. If the user turns on the app, the default filter will be set as latest history. (Newbie's filter would be checked to call all of the available users)
- Added files named `CareerData` and `ExpertiseData`. Related data will not be saved in front but sent from the server.
- There is no 'All' comparison anymore. All of the career&expertise will be saved as a number. 
- Updated related apidoc.
- Updated related unit test.

### CODE REVIEW

File: src/server/controllers/match.controller.js [hunk 1]
```diff
@@ -51,6 +62,7 @@ export function getMentorList(req, res, next) {
     findConnection(matchOptions.option1, projectOption, localField.mentee),
     findConnection(matchOptions.option2, projectOption, localField.mentor),
     findConnection(matchOptions.option3, projectOption, localField.mentee),
+    saveFilter(req.user._id, req.body.expertise, req.body.career),
```
Comment (SarahhhC) [inline src/server/controllers/match.controller.js]: If you add new promise in Promise.all, you should check result's index in next promise. Or just down that line after `User.findOne({ _id: req.user._id }).exec(),`.

File: src/server/controllers/match.controller.js [hunk 2]
```diff
@@ -121,7 +133,7 @@ export function getMentorList(req, res, next) {
 
           careerFilteredList.full.forEach((userItem) => {
             userItem.expertise.forEach((userExpertise) => {
-              if (isFitExpertiseFilter(req.body.expertise, userExpertise.select)
+              if (isArrayContainsElement(req.body.expertise, userExpertise)
```
Comment (SarahhhC) [inline src/server/controllers/match.controller.js]: Nice!

File: src/server/controllers/match.controller.js [hunk 3]
```diff
@@ -177,29 +179,38 @@ function isFitCareerFilter(userCareer, filter) {
...
-  let project = {
+  let projectOption = {
     mentee_id: 1,
     mentor_id: 1,
   };
-  let match = {
-    mentor_id: ObjectId(req.user._id),
+
+  let matchOptions = {
+    option1: {
```
Comment (SarahhhC) [inline src/server/controllers/match.controller.js]: How about name as 
`option1 => MATCH_AS_MENTOR`  
`option2 => MATCH_AS_ACCEPTED_MENTEE`  
`option3 => MATCH_AS_PENDING_MENTEE` ?

File: src/server/controllers/match.controller.js [hunk 4]
```diff
@@ -211,13 +222,16 @@ export function countExpectedExpertiseMatching(req, res, next) {
...
-    .then((mentorList) => {
+    .then((initialUserList) => {
       return new Promise((resolve, reject) => {
-        let careerFilteredList = [];
-
-        mentorList.forEach((user) => {
-          if (isFitCareerFilter(user.career, req.body.career)) {
-            careerFilteredList.push(user);
+        let careerFilteredList = {
+          full: [],
```
Comment (SarahhhC) [inline src/server/controllers/match.controller.js]: What is `full` for? I think the name is not clear.

File: src/server/controllers/match.controller.js [hunk 5]
```diff
@@ -227,9 +241,9 @@ export function countExpectedExpertiseMatching(req, res, next) {
     .then((careerFilteredList) => {
       return new Promise((resolve, reject) => {
         let countResult = [0, 0, 0, 0, 0, 0, 0];
-        careerFilteredList.forEach((user) => {
+        careerFilteredList.full.forEach((user) => {
           user.expertise.forEach((expItem) => {
-            countResult[expItem.index]++;
+            countResult[expItem - 1]++;
```
Comment (SarahhhC) [inline src/server/controllers/match.controller.js]: Is expItem same as index?

File: src/server/controllers/match.controller.js [hunk 6]
```diff
@@ -263,6 +276,34 @@ function findConnection(matchOption, projectOption, localField) {
   ]).exec();
 }
 
+function saveFilter(userId, expertise, career) {
+  return new Promise((resolve, reject) => {
+    Filter.update({
+        user_id: userId,
+        isLatest: true,
```
Comment (SarahhhC) [inline src/server/controllers/match.controller.js]: Is isLatest nessasary?? Updating would make always data latest one.

Comment (KyuaKwon) [inline src/server/controllers/match.controller.js]: I thought about some external growth about these filter history.
If we keep the history, we can gain potential insights from that.
(e.g. What kind of mentor is needed?)

File: src/server/controllers/match.controller.js (hunk 3 above)
Comment (KyuaKwon) [inline src/server/controllers/match.controller.js]: Good!

File: src/server/controllers/match.controller.js (hunk 5 above)
Comment (KyuaKwon) [inline src/server/controllers/match.controller.js]: Similar but not exactly same. Originally, the JSON structure of expItem was like this:
```
{
  index: 0,
  content: "blah",
}
```
So we counted with `index`'s value before. But we treat and compare all of the filter data (career and expertise) as a number from this PR.

Now, you'll curious why `-1` is needed. Currently, `0` means `All` in but we don't need to return the counted valued for expertise `All`. That's why.

File: src/server/controllers/match.controller.js (hunk 1 above)
Comment (KyuaKwon) [inline src/server/controllers/match.controller.js]: ok!

File: src/server/controllers/match.controller.js (hunk 6 above)
Comment (SarahhhC) [inline src/server/controllers/match.controller.js]: Good, then shouldn't we need to save filter date as well?

Comment (KyuaKwon) [inline src/server/controllers/match.controller.js]: Oh~ you're right.
I'll update it.

File: src/server/controllers/match.controller.js (hunk 4 above)
Comment (KyuaKwon) [inline src/server/controllers/match.controller.js]: `full` is an array that contains all of the information for the user object.
`_id` only contains `_id` string. This is used for checking duplication of users.

**Pull request closed without merging**